core/src/main/java/org/apache/iceberg/ManifestWriter.java

```diff
@@ -189,6 +191,18 @@ public long length() {
 
   public ManifestFile toManifestFile() {
     Preconditions.checkState(closed, "Cannot build ManifestFile, writer is not closed");
+
+    // if key metadata can store the length, add it
+    ByteBuffer keyMetadataBuffer;
+    if (keyMetadata instanceof NativeEncryptionKeyMetadata) {
+      keyMetadataBuffer =
+          ((NativeEncryptionKeyMetadata) keyMetadata).copyWithLength(length()).buffer();
```

Comment:
Updated this to add the length to the encryption metadata, since it should now be stored there.